pkg/messaging/topics.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Abstract Machines
+// SPDX-License-Identifier: Apache-2.0
+
+package messaging
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/absmach/supermq/pkg/errors"
+)
+
+const (
+	MsgTopicPrefix     = "m"
+	ChannelTopicPrefix = "c"
+
+	numGroups     = 4 // entire expression + domain group + channel group + subtopic group
+	domainGroup   = 1 // domain group is first in msg topic regexp
+	channelGroup  = 2 // channel group is second in msg topic regexp
+	subtopicGroup = 3 // subtopic group is third in msg topic regexp
+)
+
+var (
+	ErrMalformedTopic    = errors.New("malformed topic")
+	ErrMalformedSubtopic = errors.New("malformed subtopic")
+	// Regex to group topic in format m.<domain_id>.c.<channel_id>.<sub_topic> `^\/?m\/([\w\-]+)\/c\/([\w\-]+)(\/[^?]*)?(\?.*)?$`.
+	topicRegExp          = regexp.MustCompile(`^\/?` + MsgTopicPrefix + `\/([\w\-]+)\/` + ChannelTopicPrefix + `\/([\w\-]+)(\/[^?]*)?(\?.*)?$`)
```

Comment:
Regex is slow, as you can see from the attached benchmarks. Since it is not replaceable in our case, let's not use it. We need to write our parser, and we need to be careful, so let's do it in a separate PR.